File: Queue/Linked-List-Based-implementation/CPP/linkedList.cpp
#include <iostream>
using namespace std;

class linkedListNode {
public:
    int data;
    linkedListNode *next;
    linkedListNode(int _data) {
        this->data = _data;
        this->next = NULL;
    }
};

class linkedListIterator {
private:
    linkedListNode *currentNode;

public:
    linkedListIterator() { currentNode = NULL; }
    linkedListIterator(linkedListNode *node) { currentNode = node; }
    int data() { return currentNode->data; }
    linkedListIterator next() {
        currentNode = currentNode->next;
        return *this;
    }

    linkedListNode *current() { return currentNode; }
};

class linkedList {
private:
    int length;

public:
    linkedListNode *head = NULL;
    linkedListNode *tail = NULL;

    void addHead(linkedListNode *node) {
        this->head = node;
        this->tail = node;
    }

    void insertLast(int _data) {
        linkedListNode *newNode = new linkedListNode(_data);
        if (this->head == NULL) {
            this->head = newNode;
            this->tail = newNode;
        } else {
            this->tail->next = newNode;
            this->tail = newNode;
        }
        this->length++;
    }

    void insertFirst(int _data) {
        linkedListNode *newNode = new linkedListNode(_data);
        if (this->head == NULL) {
            this->head = newNode;
            this->tail = newNode;
        } else {
            newNode->next = this->head;
            this->head = newNode;
        }
        this->length++;
    }

    void insertAfter(linkedListNode *node, int _data) {
        linkedListNode *newnode = new linkedListNode(_data);
        newnode->next = node->next;
        node->next = newnode;
        if (newnode->next == NULL) {
            this->tail = newnode;
        }
        this->length++;
    }

    void deleteNode(linkedListNode *node) {
        if (this->head == this->tail) {
            this->head = NULL;
            this->tail = NULL;
        } else if (this->head == node) {
            this->head = node->next;
        } else {
            linkedListNode *parentNode = this->findParent(node);
            if (this->tail == node) {
                this->tail = parentNode;
            } else {
                parentNode->next = node->next;
            }
        }
        delete node;
        this->length--;
    }

    void deleteNode(int _data) {
        linkedListNode *node = this->find(_data);
        if (node == NULL) {
            return;
        }
        this->deleteNode(node);
    }

    void deleteHead() {
        if (this->head == NULL) {
            return;
        }
        this->deleteNode(this->head);
    }

    void insertBefore(linkedListNode *node, int _data) {
        linkedListNode *newnode = new linkedListNode(_data);
        newnode->next = node;

        linkedListNode *parentNode = this->findParent(node);

        if (parentNode == NULL) {
            this->head = newnode;
        } else {
            parentNode->next = newnode;
        }
        this->length++;
    }

    linkedListNode *find(int _data) {
        for (auto itr = this->begin(); itr.current() != NULL; itr.next()) {
            if (itr.data() == _data) {
                return itr.current();
            }
        }
        return NULL;
    }

    linkedListNode *findParent(linkedListNode *node) {
        for (auto itr = this->begin(); itr.current() != NULL; itr.next()) {
            if (itr.current()->next == node) {
                return itr.current();
            }
        }
        return NULL;
    }

    int getLengthItr() {
        int count = 0;
        for (auto itr = this->begin(); itr.current() != NULL; itr.next()) {
            count++;
        }
        return count;
    }

    int Length() { return this->length; }

    void printList() {
        for (auto itr = this->begin(); itr.current() != NULL; itr.next()) {
            cout << itr.data() << " -> ";
        }
        cout << "\n";
    }

    int sum() {
        int sum = 0;
        for (auto itr = this->begin(); itr.current() != NULL; itr.next()) {
            sum += itr.data();
        }
        return sum;
    }

    linkedListIterator begin() {
        linkedListIterator itr(this->head);
        return itr;
    }
};

File: Stack/Linked-list-base-implementation/LinkedList.js
class LinkedListNode {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedListIterator {
    constructor(head) {
        this.currentNode = head;
    }

    data() {
        return this.currentNode ? this.currentNode.data : null;
    }

    next() {
        if (this.currentNode) {
            this.currentNode = this.currentNode.next;
        }
    }

    current() {
        return this.currentNode;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
        this.uniqueValues = new Set();
    }

    begin() {
        return new LinkedListIterator(this.head);
    }

    printList() {
        let current = this.head;
        while (current) {
            console.log(current.data);
            current = current.next;
        }
    }

    find(data) {
        let current = this.head;
        while (current) {
            if (current.data === data) {
                return current;
            }
            current = current.next;
        }
        return null;
    }

    findParent(node) {
        let current = this.head;
        while (current && current.next !== node) {
            current = current.next;
        }
        return current;
    }

    insertLast(data) {
        const newNode = new LinkedListNode(data);
        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
        this.uniqueValues.add(data);
    }

    insertAfter(existingData, newData) {
        const existingNode = this.find(existingData);
        if (existingNode) {
            const newNode = new LinkedListNode(newData);
            newNode.next = existingNode.next;
            existingNode.next = newNode;
            if (this.tail === existingNode) {
                this.tail = newNode;
            }
            this.length++;
            this.uniqueValues.add(newData);
        }
    }

    insertFirst(data) {
        if (!this.canInsert(data)) return;
        const newNode = new LinkedListNode(data);  // Corrected the class name
        if (this.head == null) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }
        this.length++;
        this.uniqueValues.add(data);  // Ensure to track the new node in uniqueValues
    }

    insertBefore(existingData, newData) {
        if (this.head && this.head.data === existingData) {
            const newNode = new LinkedListNode(newData);
            newNode.next = this.head;
            this.head = newNode;
            this.length++;
            this.uniqueValues.add(newData);
            return;
        }
        const existingNode = this.find(existingData);
        if (existingNode) {
            const parentNode = this.findParent(existingNode);
            const newNode = new LinkedListNode(newData);
            newNode.next = existingNode;
            parentNode.next = newNode;
            this.length++;
            this.uniqueValues.add(newData);
        }
    }

    deleteNode(data) {
        if (!this.head) return;

        if (this.head.data === data) {
            this.head = this.head.next;
            this.length--;
            this.uniqueValues.delete(data);
            return;
        }

        let current = this.head;
        while (current.next) {
            if (current.next.data === data) {
                current.next = current.next.next;
                if (current.next === null) {
                    this.tail = current;
                }
                this.length--;
                this.uniqueValues.delete(data);
                return;
            }
            current = current.next;
        }
    }

    deleteHead() {
        if (this.head == null) return;
        this.head = this.head.next;
        this.length--;
    }

    isExist(data) {
        return this.uniqueValues.has(data);
    }

    canInsert(data) {
        return !this.isExist(data);
    }
}



module.exports = LinkedList;
